# Data Wrangling Project - WeRateDogs Twitter Archive

## Introduction
WeRateDogs is a Twitter account that rates people's dogs with a humorous comment about the dog. These ratings almost always have a denominator of 10. The numerators, though? Almost always greater than 10. 11/10, 12/10, 13/10, etc. Why? Because "they're good dogs Brent." WeRateDogs has over 4 million followers and has received international media coverage.

WeRateDogs downloaded their Twitter archive and sent it to Udacity via email exclusively for you to use in this project. This archive contains basic tweet data (tweet ID, timestamp, text, etc.) for all 5000+ of their tweets as they stood on August 1, 2017.

![WeRateDogs Twitter](https://video.udacity-data.com/topher/2017/October/59dd378f_dog-rates-social/dog-rates-social.jpg)

### Project Steps Overview
With the main focus of this project being on data wrangling, it will be divided into the following steps:
1. Gathering data
2. Assessing data
3. Cleaning data
4. Storing data
5. Analyzing and visualizing data
6. Reporting
      - My data wrangling efforts
      - My data analyses and visualizations.

### Aim
The goal is to create interesting and trustworthy analyses and visualizations. The Twitter archive is great, but it only contains very basic tweet information. Additional gathering, then assessing and cleaning is required for "Wow!"-worthy analyses and visualizations.

### The Data
In this project, I will work on the following three datasets.

##### Enhanced Twitter Archive
This data was provided by the Udacity team. The WeRateDogs Twitter archive contains basic tweet data for all 5000+ of their tweets, but not everything. One column the archive does contain though: each tweet's text, which was used to extract rating, dog name, and dog "stage" (i.e. doggo, floofer, pupper, and puppo) to make this Twitter archive "enhanced." Of the 5000+ tweets, they filtered for tweets with ratings only (there are 2356).
![image.png](https://video.udacity-data.com/topher/2017/October/59dd4791_screenshot-2017-10-10-18.19.36/screenshot-2017-10-10-18.19.36.png)

They extracted this data programmatically, but they didn't do a very good job. The ratings probably aren't all correct. Same goes for the dog names and probably dog stages (see below for more information on these) too. I'll need to assess and clean these columns if I want to use them for analysis and visualization.

##### Additional Data via the Twitter API
Back to the basic-ness of Twitter archives: ***retweet count*** and ***favorite count*** are two of the notable column omissions. Fortunately, this additional data can be gathered by anyone from Twitter's API. Since I have access to Twitter's API and the tweet IDs of the tweets in the Enhanced archive, I can gather the needed data for all 5000+ by querying Twitter's API.

##### Image Predictions File
Every image in the WeRateDogs Twitter archive was run through a neural network that can classify breeds of dogs. The results: a table full of image predictions (the top three only) alongside each tweet ID, image URL, and the image number that corresponded to the most confident prediction (numbered 1 to 4 since tweets can have up to four images).
![imgpredict](https://video.udacity-data.com/topher/2017/October/59dd4d2c_screenshot-2017-10-10-18.43.41/screenshot-2017-10-10-18.43.41.png)

So for the last row in that table:
* `tweet_id` is the last part of the tweet URL after "status/" → https://twitter.com/dog_rates/status/889531135344209921
* `p1` is the algorithm's #1 prediction for the image in the tweet → **golden retriever**
* `p1_conf` is how confident the algorithm is in its #1 prediction → **95%**
* `p1_dog` is whether or not the #1 prediction is a breed of dog → **TRUE**
* `p2` is the algorithm's second most likely prediction → **Labrador retriever**
* `p2_conf` is how confident the algorithm is in its #2 prediction → **1%**
* `p2_dog` is whether or not the #2 prediction is a breed of dog → **TRUE**
* etc.

## Gathering Data

In [2]:
#%pip install tweepy

In [1]:
import pandas as pd
import numpy as np
import requests
import json
import os

### Data downloaded manually
The WeRateDogs Twitter archive was given to us by hand, so let's read it into a pandas dataframe.

In [3]:
twitter_enhanced_df = pd.read_csv('twitter-archive-enhanced.csv')

In [4]:
twitter_enhanced_df.head()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   
3  891689557279858688                    NaN                  NaN   
4  891327558926688256                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   
3  2017-07-30 15:58:51 +0000   
4  2017-07-29 16:00:24 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3  This is Darla. She commenced a snooze mid meal...                  NaN   
4  This is Franklin. He would like you to stop ca...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   
3                       NaN                        NaN   
4                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   
3  https://twitter.com/dog_rates/status/891689557...                13   
4  https://twitter.com/dog_rates/status/891327558...                12   

   rating_denominator      name doggo floofer pupper puppo  
0                  10   Phineas  None    None   None  None  
1                  10     Tilly  None    None   None  None  
2                  10    Archie  None    None   None  None  
3                  10     Darla  None    None   None  None  
4                  10  Franklin  None    None   None  None

### Data from Web
Downloading the **tweet image predictions** programmatically using the `requests` library.

In [5]:
url = "https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv"
response = requests.get(url)

In [15]:
with open("image_predictions.tsv",mode='wb') as file:
      file.write(response.content)

In [16]:
os.listdir()

['image_predictions.tsv',
 'twitter-archive-enhanced.csv',
 'WeRateDogs Twitter_Data Wrangling Project.ipynb']

In [18]:
img_predictions_df = pd.read_csv('image_predictions.tsv', sep='\t')

In [19]:
img_predictions_df.head()

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   
3        1     Rhodesian_ridgeback  0.408143    True             redbone   
4        1      miniature_pinscher  0.560311    True          Rottweiler   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True  
2  0.138584    True           bloodhound  0.116197    True  
3  0.360687    True   miniature_pinscher  0.222752    True  
4  0.243682    True             Doberman  0.154629    True

### Data from API
Use Tweepy to query Twitter's API for the retweet count and favorite count for each tweet.

- Create API object
- Create empty list for containing dictionary with keys 'tweet_id' and 'json_string' or 'tweet_data'
- Create an empty list for storing errors
- Iterate through the tweet_id in the twitter_enhanced_df
- Use the try-except statement to catch errors
- Monitor each iterations time.
- Get status of tweet
- Convert json status to string so that it can be stored.
- Store the tweet_id and json_string in a dictionary
- Store dictionary in empty list
- Store errors in list.
- Work on the errors
- Store each json_string in a new line in `tweet_json.txt` file

In [71]:
import tweepy
import timeit

My API keys, secrets and tokens won't be included this report. You can get yours by signing up for the [Twitter Developer Account](https://developer.twitter.com)

In [27]:
api_key = "your api key"
api_key_secret = "your api key secret"
access_token = "your access token"
access_token_secret = "your access token secret"

Creating the API object that I'll use to gather Twitter data.

In [36]:
#auth = tweepy.OAuthHandler(api_key, api_key_secret)
#auth.set_access_token(access_token, access_token_secret)

api = tweepy.API(auth=auth, wait_on_rate_limit=True)

Defining relevant functions

In [74]:
def tweet_status(tweetID):
      '''
      To get the status of a particular tweetID. Returns a dictioary of tweet ID and
      the json data in string format.
      '''
      # get status of the tweet
      status = api.get_status(tweetID, tweet_mode='extended')
      # get the json data in string
      tweet_data = json.dumps(status._json)
      
      return {'tweet_id': tweetID, 'tweet_data':tweet_data}

In [83]:
'''
tweets = [] # creating empty list for containing dictionary returned from tweet_status()
errorList = [] # list for containing the failed tweet IDs

i = 1 # counter
for id in twitter_enhanced_df.tweet_id:
      
      print("{}: {}".format(i, id))
      
      start = timeit.timeit() # get start time
      try:
            status = tweet_status(id) # get status for tweet ID
            tweets.append(status) # append return dictionary to tweet_json
      except Exception as e:
            # get the tweet ID and its particular error
            errorList.append({
                              'tweet_id': id,
                              'error': str(e)
                              })
      end = timeit.timeit() # get end time
      print("\truntime:",end - start)
      i += 1
'''

1: 892420643555336193
	runtime: 0.03284029999849736
2: 892177421306343426
	runtime: 0.016495300002134172
3: 891815181378084864
	runtime: 0.00876839999909862
4: 891689557279858688
	runtime: 0.007181000004493399
5: 891327558926688256
	runtime: 0.011067400002502836
6: 891087950875897856
	runtime: 0.005447499999718275
7: 890971913173991426
	runtime: 0.01020059999791556
8: 890729181411237888
	runtime: 0.014640300003520679
9: 890609185150312448
	runtime: 0.007319699998333817
10: 890240255349198849
	runtime: 0.009109599999646889
11: 890006608113172480
	runtime: -0.0013445999975374434
12: 889880896479866881
	runtime: 0.00751259999742615
13: 889665388333682689
	runtime: 0.012810500000341563
14: 889638837579907072
	runtime: 0.010224100002233172
15: 889531135344209921
	runtime: 0.009715500000311295
16: 889278841981685760
	runtime: 0.009460100005526328
17: 888917238123831296
	runtime: 0.016081899997516302
18: 888804989199671297
	runtime: 0.0034466999968572054
19: 888554962724278272
	runtime: 0.012

Rate limit reached. Sleeping for: 178


	runtime: 0.012303399998927489
891: 759566828574212096
	runtime: -0.0005553000009967946
892: 759557299618865152
	runtime: 0.004020400003355462
893: 759447681597108224
	runtime: 0.009658200000558281
894: 759446261539934208
	runtime: 0.015201199999864912
895: 759197388317847553
	runtime: -0.012521699998615077
896: 759159934323924993
	runtime: 0.0023095000033208635
897: 759099523532779520
	runtime: 0.0030828999988443684
898: 759047813560868866
	runtime: 0.008664800003316486
899: 758854675097526272
	runtime: 0.008317700001498451
900: 758828659922702336
	runtime: 0.005359300001146039
901: 758740312047005698
	runtime: 0.012296600001718616
902: 758474966123810816
	runtime: -0.005912199998419965
903: 758467244762497024
	runtime: 0.014805300001171418
904: 758405701903519748
	runtime: -0.005016799997974886
905: 758355060040593408
	runtime: 0.05482090000077733
906: 758099635764359168
	runtime: -0.002992199999425793
907: 758041019896193024
	runtime: 0.01646710000204621
908: 757741869644341248
	run

Rate limit reached. Sleeping for: 220


	runtime: 0.009938099999999395
1791: 677530072887205888
	runtime: 0.006335800000670133
1792: 677335745548390400
	runtime: 0.005997300002491102
1793: 677334615166730240
	runtime: 0.007152399997721659
1794: 677331501395156992
	runtime: 0.006478399998741224
1795: 677328882937298944
	runtime: 0.0038512000028276816
1796: 677314812125323265
	runtime: 0.009676499998022337
1797: 677301033169788928
	runtime: 0.005042700006015366
1798: 677269281705472000
	runtime: 0.007750399996439228
1799: 677228873407442944
	runtime: 0.013334999999642605
1800: 677187300187611136
	runtime: -0.0013901000020268839
1801: 676975532580409345
	runtime: 0.0058262999991711695
1802: 676957860086095872
	runtime: -0.0006466000013460871
1803: 676949632774234114
	runtime: 8.789999992586672e-05
1804: 676948236477857792
	runtime: 0.009919500000250991
1805: 676946864479084545
	runtime: -0.0027098999998997897
1806: 676942428000112642
	runtime: 0.00357570000414853
1807: 676936541936185344
	runtime: 0.020049400001880713
1808: 676

In [115]:
errorList.__len__()

29

In [119]:
pd.DataFrame(errorList)

tweet_id                                              error
0   888202515573088257  404 Not Found\n144 - No status found with that...
1   873697596434513921  404 Not Found\n144 - No status found with that...
2   872668790621863937  404 Not Found\n144 - No status found with that...
3   872261713294495745  404 Not Found\n144 - No status found with that...
4   869988702071779329  404 Not Found\n144 - No status found with that...
5   866816280283807744  404 Not Found\n144 - No status found with that...
6   861769973181624320  404 Not Found\n144 - No status found with that...
7   856602993587888130  404 Not Found\n144 - No status found with that...
8   856330835276025856  404 Not Found\n144 - No status found with that...
9   851953902622658560  404 Not Found\n144 - No status found with that...
10  851861385021730816  404 Not Found\n144 - No status found with that...
11  845459076796616705  404 Not Found\n144 - No status found with that...
12  844704788403113984  404 Not Found\n144 - No status found with that...
13  842892208864923648  404 Not Found\n144 - No status found with that...
14  837366284874571778  404 Not Found\n144 - No status found with that...
15  837012587749474308  404 Not Found\n144 - No status found with that...
16  829374341691346946  404 Not Found\n144 - No status found with that...
17  827228250799742977  404 Not Found\n144 - No status found with that...
18  812747805718642688  404 Not Found\n144 - No status found with that...
19  802247111496568832  404 Not Found\n144 - No status found with that...
20  779123168116150273  404 Not Found\n144 - No status found with that...
21  775096608509886464  404 Not Found\n144 - No status found with that...
22  771004394259247104  403 Forbidden\n179 - Sorry, you are not author...
23  770743923962707968  404 Not Found\n144 - No status found with that...
24  766864461642756096  404 Not Found\n144 - No status found with that...
25  759923798737051648  404 Not Found\n144 - No status found with that...
26  759566828574212096  404 Not Found\n144 - No status found with that...
27  754011816964026368  404 Not Found\n144 - No status found with that...
28  680055455951884288  404 Not Found\n144 - No status found with that...

Looks like we couldn't get the tweet data for 29 out of the 2356 tweet IDs that we have. Further investigation suggests that these tweets were deleted (`404` error) or the user's account is private (`403` error).

Let's write the successful json tweet data into a file `tweet_json.txt`.

In [123]:
tweets[0]

{'tweet_id': 892420643555336193,
 'tweet_data': '{"created_at": "Tue Aug 01 16:23:56 +0000 2017", "id": 892420643555336193, "id_str": "892420643555336193", "full_text": "This is Phineas. He\'s a mystical boy. Only ever appears in the hole of a donut. 13/10 https://t.co/MgUWQ76dJU", "truncated": false, "display_text_range": [0, 85], "entities": {"hashtags": [], "symbols": [], "user_mentions": [], "urls": [], "media": [{"id": 892420639486877696, "id_str": "892420639486877696", "indices": [86, 109], "media_url": "http://pbs.twimg.com/media/DGKD1-bXoAAIAUK.jpg", "media_url_https": "https://pbs.twimg.com/media/DGKD1-bXoAAIAUK.jpg", "url": "https://t.co/MgUWQ76dJU", "display_url": "pic.twitter.com/MgUWQ76dJU", "expanded_url": "https://twitter.com/dog_rates/status/892420643555336193/photo/1", "type": "photo", "sizes": {"thumb": {"w": 150, "h": 150, "resize": "crop"}, "medium": {"w": 540, "h": 528, "resize": "fit"}, "small": {"w": 540, "h": 528, "resize": "fit"}, "large": {"w": 540, "h": 528, 

In [140]:
# write only 'tweet_data' of each dictionary in `tweets` to a txt file
c = 0
for status in tweets:
      # creating file and storing data
      if c == 0:
            with open('tweet_json.txt', 'w') as file:
                  file.write(status['tweet_data'])
      else: # appending each json data as a new line
            with open('tweet_json.txt', 'a') as file:
                  file.write("\n"+status['tweet_data'])
      c = 1

Read `tweet_json.txt` line by line into a pandas DataFrame with `tweet ID`, `retweet count` and `favorite count`.

In [152]:
tweet_json = [] # creating empty list to store dictionary

with open('tweet_json.txt', 'r') as file:
      # iterating through each line in 'file'
      for line in file.readlines():
            # converting 'line' from string to dictionary type
            data = json.loads(line)
            # appending dictionary contain our values of interest to 'tweet_json'
            tweet_json.append({
                              'tweet_id': data['id'],
                              'retweet_count': data['retweet_count'],
                              'favorite_count': data['favorite_count']
                               })

In [154]:
tweet_json_df = pd.DataFrame(tweet_json)

In [155]:
tweet_json_df

tweet_id  retweet_count  favorite_count
0     892420643555336193           6969           33693
1     892177421306343426           5272           29218
2     891815181378084864           3464           21974
3     891689557279858688           7191           36785
4     891327558926688256           7715           35178
...                  ...            ...             ...
2322  666049248165822465             36              88
2323  666044226329800704            115             246
2324  666033412701032449             36             100
2325  666029285002620928             39             112
2326  666020888022790149            419            2282

[2327 rows x 3 columns]

## Assessing Data
### Visual Assessment
* **Quality Issues**

`twitter_enhanced_df`
1.  Some names in `name` column don't have the correct name
2.  Some rows in  have a 'None' value for `doggo`, `floofer`, `pupper` and `puppo` columns.
3. `in_reply_to_status_id`, `in_reply_to_user_id`, `retweeted_status_id`, `retweeted_status_user_id` and `retweeted_status_timestamp` have NaN values.

`img_predictions_df`
1. For some Twitter IDs, `p1_dog`, `p2_dog` and `p3_dog` are all False indicating that there's no correct dog breed prediction for those IDs.
2. Names of dog breeds in `p1`, `p2` and `p3` are in lowercases.

`tweet_json_df`
1. 29 tweet IDs from `twitter_enhanced_df` don't have a record here.

In [161]:
twitter_enhanced_df.head(10)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   
3  891689557279858688                    NaN                  NaN   
4  891327558926688256                    NaN                  NaN   
5  891087950875897856                    NaN                  NaN   
6  890971913173991426                    NaN                  NaN   
7  890729181411237888                    NaN                  NaN   
8  890609185150312448                    NaN                  NaN   
9  890240255349198849                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   
3  2017-07-30 15:58:51 +0000   
4  2017-07-29 16:00:24 +0000   
5  2017-07-29 00:08:17 +0000   
6  2017-07-28 16:27:12 +0000   
7  2017-07-28 00:22:40 +0000   
8  2017-07-27 16:25:51 +0000   
9  2017-07-26 15:59:51 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   
5  <a href="http://twitter.com/download/iphone" r...   
6  <a href="http://twitter.com/download/iphone" r...   
7  <a href="http://twitter.com/download/iphone" r...   
8  <a href="http://twitter.com/download/iphone" r...   
9  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3  This is Darla. She commenced a snooze mid meal...                  NaN   
4  This is Franklin. He would like you to stop ca...                  NaN   
5  Here we have a majestic great white breaching ...                  NaN   
6  Meet Jax. He enjoys ice cream so much he gets ...                  NaN   
7  When you watch your owner call another dog a g...                  NaN   
8  This is Zoey. She doesn't want to be one of th...                  NaN   
9  This is Cassie. She is a college pup. Studying...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   
3                       NaN                        NaN   
4                       NaN                        NaN   
5                       NaN                        NaN   
6                       NaN                        NaN   
7                       NaN                        NaN   
8                       NaN                        NaN   
9                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   
3  https://twitter.com/dog_rates/status/891689557...                13   
4  https://twitter.com/dog_rates/status/891327558...                12   
5  https://twitter.com/dog_rates/status/891087950...                13   
6  https://gofundme.com/ydvmve-surgery-for-jax,ht...                13   
7  https://twitter.com/dog_rates/status/890729181...                13   
8  https://twitter.com/dog_rates/status/890609185...                13   
9  https://twitter.com/dog_rates/status/890240255...    

In [165]:
twitter_enhanced_df.tail()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
2351  666049248165822465                    NaN                  NaN   
2352  666044226329800704                    NaN                  NaN   
2353  666033412701032449                    NaN                  NaN   
2354  666029285002620928                    NaN                  NaN   
2355  666020888022790149                    NaN                  NaN   

                      timestamp  \
2351  2015-11-16 00:24:50 +0000   
2352  2015-11-16 00:04:52 +0000   
2353  2015-11-15 23:21:54 +0000   
2354  2015-11-15 23:05:30 +0000   
2355  2015-11-15 22:32:08 +0000   

                                                 source  \
2351  <a href="http://twitter.com/download/iphone" r...   
2352  <a href="http://twitter.com/download/iphone" r...   
2353  <a href="http://twitter.com/download/iphone" r...   
2354  <a href="http://twitter.com/download/iphone" r...   
2355  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
2351  Here we have a 1949 1st generation vulpix. Enj...                  NaN   
2352  This is a purebred Piers Morgan. Loves to Netf...                  NaN   
2353  Here is a very happy pup. Big fan of well-main...                  NaN   
2354  This is a western brown Mitsubishi terrier. Up...                  NaN   
2355  Here we have a Japanese Irish Setter. Lost eye...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
2351                       NaN                        NaN   
2352                       NaN                        NaN   
2353                       NaN                        NaN   
2354                       NaN                        NaN   
2355                       NaN                        NaN   

                                          expanded_urls  rating_numerator  \
2351  https://twitter.com/dog_rates/status/666049248...                 5   
2352  https://twitter.com/dog_rates/status/666044226...                 6   
2353  https://twitter.com/dog_rates/status/666033412...                 9   
2354  https://twitter.com/dog_rates/status/666029285...                 7   
2355  https://twitter.com/dog_rates/status/666020888...                 8   

      rating_denominator  name doggo floofer pupper puppo  
2351                  10  None  None    None   None  None  
2352                  10     a  None    None   None  None  
2353                  10     a  None    None   None  None  
2354                  10     a  None    None   None  None  
2355                  10  None  None    None   None  None

In [168]:
twitter_enhanced_df.sample(15)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
1855  675531475945709568                    NaN                  NaN   
2241  667915453470232577                    NaN                  NaN   
1287  708356463048204288                    NaN                  NaN   
1791  677335745548390400                    NaN                  NaN   
1715  680221482581123072                    NaN                  NaN   
819   770655142660169732                    NaN                  NaN   
1895  674742531037511680           6.747400e+17         4.196984e+09   
2198  668815180734689280                    NaN                  NaN   
632   793962221541933056                    NaN                  NaN   
1604  685906723014619143                    NaN                  NaN   
939   753039830821511168                    NaN                  NaN   
1520  690728923253055490                    NaN                  NaN   
1213  715220193576927233                    NaN                  NaN   
1779  677716515794329600                    NaN                  NaN   
466   817171292965273600                    NaN                  NaN   

                      timestamp  \
1855  2015-12-12 04:23:49 +0000   
2241  2015-11-21 04:00:28 +0000   
1287  2016-03-11 18:18:36 +0000   
1791  2015-12-17 03:53:20 +0000   
1715  2015-12-25 03:00:14 +0000   
819   2016-08-30 16:11:18 +0000   
1895  2015-12-10 00:08:50 +0000   
2198  2015-11-23 15:35:39 +0000   
632   2016-11-02 23:45:19 +0000   
1604  2016-01-09 19:31:20 +0000   
939   2016-07-13 01:34:21 +0000   
1520  2016-01-23 02:53:03 +0000   
1213  2016-03-30 16:52:36 +0000   
1779  2015-12-18 05:06:23 +0000   
466   2017-01-06 00:49:53 +0000   

                                                 source  \
1855  <a href="http://twitter.com/download/iphone" r...   
2241  <a href="http://twitter.com/download/iphone" r...   
1287  <a href="http://twitter.com/download/iphone" r...   
1791  <a href="http://vine.co" rel="nofollow">Vine -...   
1715  <a href="http://twitter.com/download/iphone" r...   
819   <a href="http://twitter.com/download/iphone" r...   
1895  <a href="http://twitter.com/download/iphone" r...   
2198  <a href="http://twitter.com/download/iphone" r...   
632   <a href="http://twitter.com/download/iphone" r...   
1604  <a href="http://twitter.com/download/iphone" r...   
939   <a href="http://vine.co" rel="nofollow">Vine -...   
1520  <a href="http://twitter.com/download/iphone" r...   
1213  <a href="http://twitter.com/download/iphone" r...   
1779  <a href="http://twitter.com/download/iphone" r...   
466   <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
1855  This is Ellie AKA Queen Slayer of the Orbs. Ve...                  NaN   
2241  Meet Otis. He is a Peruvian Quartzite. Pic spo...                  NaN   
1287  This is Oliver. That is his castle. He protect...                  NaN   
1791  Downright inspiring 12/10 https://t.co/vSLtYBWHcQ                  NaN   
1715  This is CeCe. She's patiently waiting for Sant...                  NaN   
819   We only rate dogs. Pls stop sending in non-can...                  NaN   
1895  Some clarification is required. The dog is sin...                  NaN   
2198  This is a wild Toblerone from Papua New Guinea...                  NaN   
632   This is Maximus. His face is stuck like that. ...                  NaN   
1604  This is Olive. He's stuck in a sleeve. 9/10 da...                  NaN   
939   So this just changed my life. 13/10 please enj...                  NaN   
1520  This is Vinscent. He was just questioned about...                  NaN   
1213  This is Nico. His selfie game is strong af. Ex...                  NaN   
1779  IT'S PUPPERGEDDON. Total of 144/120 ...I think...                  NaN   
466   This is Tebow. He kindly requests that you put...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
1855           

In [166]:
img_predictions_df.head(15)

tweet_id                                          jpg_url  \
0   666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1   666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2   666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3   666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4   666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   
5   666050758794694657  https://pbs.twimg.com/media/CT5Jof1WUAEuVxN.jpg   
6   666051853826850816  https://pbs.twimg.com/media/CT5KoJ1WoAAJash.jpg   
7   666055525042405380  https://pbs.twimg.com/media/CT5N9tpXIAAifs1.jpg   
8   666057090499244032  https://pbs.twimg.com/media/CT5PY90WoAAQGLo.jpg   
9   666058600524156928  https://pbs.twimg.com/media/CT5Qw94XAAA_2dP.jpg   
10  666063827256086533  https://pbs.twimg.com/media/CT5Vg_wXIAAXfnj.jpg   
11  666071193221509120  https://pbs.twimg.com/media/CT5cN_3WEAAlOoZ.jpg   
12  666073100786774016  https://pbs.twimg.com/media/CT5d9DZXAAALcwe.jpg   
13  666082916733198337  https://pbs.twimg.com/media/CT5m4VGWEAAtKc8.jpg   
14  666094000022159362  https://pbs.twimg.com/media/CT5w9gUW4AAsBNN.jpg   

    img_num                      p1   p1_conf  p1_dog                  p2  \
0         1  Welsh_springer_spaniel  0.465074    True              collie   
1         1                 redbone  0.506826    True  miniature_pinscher   
2         1         German_shepherd  0.596461    True            malinois   
3         1     Rhodesian_ridgeback  0.408143    True             redbone   
4         1      miniature_pinscher  0.560311    True          Rottweiler   
5         1    Bernese_mountain_dog  0.651137    True    English_springer   
6         1              box_turtle  0.933012   False          mud_turtle   
7         1                    chow  0.692517    True     Tibetan_mastiff   
8         1           shopping_cart  0.962465   False     shopping_basket   
9         1        miniature_poodle  0.201493    True            komondor   
10        1        golden_retriever  0.775930    True     Tibetan_mastiff   
11        1           Gordon_setter  0.503672    True   Yorkshire_terrier   
12        1            Walker_hound  0.260857    True    English_foxhound   
13        1                     pug  0.489814    True        bull_mastiff   
14        1              bloodhound  0.195217    True     German_shepherd   

     p2_conf  p2_dog                           p3   p3_conf  p3_dog  
0   0.156665    True            Shetland_sheepdog  0.061428    True  
1   0.074192    True          Rhodesian_ridgeback  0.072010    True  
2   0.138584    True                   bloodhound  0.116197    True  
3   0.360687    True           miniature_pinscher  0.222752    True  
4   0.243682    True                     Doberman  0.154629    True  
5   0.263788    True   Greater_Swiss_Mountain_dog  0.016199    True  
6   0.045885   False                     terrapin  0.017885   False  
7   0.058279    True                     fur_coat  0.054449   False  
8   0.014594   False             golden_retriever  0.007959    True  
9   0.192305    True  soft-coated_wheaten_terrier  0.082086    True  
10  0.093718    True           Labrador_retriever  0.072427    True  
11  0.174201    True                     Pekinese  0.109454    True  
12  0.175382    True                 Ibizan_hound  0.097471    True  
13  0.404722    True               French_bulldog  0.048960    True  
14  0.078260    True                     malinois  0.075628    True

In [169]:
img_predictions_df.tail()

tweet_id                                          jpg_url  \
2070  891327558926688256  https://pbs.twimg.com/media/DF6hr6BUMAAzZgT.jpg   
2071  891689557279858688  https://pbs.twimg.com/media/DF_q7IAWsAEuuN8.jpg   
2072  891815181378084864  https://pbs.twimg.com/media/DGBdLU1WsAANxJ9.jpg   
2073  892177421306343426  https://pbs.twimg.com/media/DGGmoV4XsAAUL6n.jpg   
2074  892420643555336193  https://pbs.twimg.com/media/DGKD1-bXoAAIAUK.jpg   

      img_num           p1   p1_conf  p1_dog                  p2   p2_conf  \
2070        2       basset  0.555712    True    English_springer  0.225770   
2071        1  paper_towel  0.170278   False  Labrador_retriever  0.168086   
2072        1    Chihuahua  0.716012    True            malamute  0.078253   
2073        1    Chihuahua  0.323581    True            Pekinese  0.090647   
2074        1       orange  0.097049   False               bagel  0.085851   

      p2_dog                           p3   p3_conf  p3_dog  
2070    True  German_short-haired_pointer  0.175219    True  
2071    True                      spatula  0.040836   False  
2072    True                       kelpie  0.031379    True  
2073    True                     papillon  0.068957    True  
2074   False                       banana  0.076110   False

In [189]:
img_predictions_df.sample(15)

tweet_id                                            jpg_url  \
274   670833812859932673    https://pbs.twimg.com/media/CU9HyzSWIAAVcte.jpg   
521   676575501977128964    https://pbs.twimg.com/media/CWOt07EUsAAnOYW.jpg   
677   683481228088049664    https://pbs.twimg.com/media/CXw2jSpWMAAad6V.jpg   
1777  828381636999917570    https://pbs.twimg.com/media/C38Asz1WEAAvzj3.jpg   
1304  753294487569522689    https://pbs.twimg.com/media/CnQ9Vq1WEAEYP01.jpg   
827   693486665285931008  https://pbs.twimg.com/ext_tw_video_thumb/69348...   
945   704480331685040129    https://pbs.twimg.com/media/CcbRIAgXIAQaKHQ.jpg   
289   671163268581498880    https://pbs.twimg.com/media/CVBzbWsWsAEyNMA.jpg   
304   671518598289059840    https://pbs.twimg.com/media/CVG2l9jUYAAwg-w.jpg   
1247  747600769478692864    https://pbs.twimg.com/media/CmAC7ehXEAAqSuW.jpg   
938   703631701117943808    https://pbs.twimg.com/media/CcPNS4yW8AAd-Et.jpg   
1586  797971864723324932    https://pbs.twimg.com/media/CxL3IWeVEAAAIE2.jpg   
990   707995814724026368    https://pbs.twimg.com/media/CdNOb17WwAA5z4A.jpg   
1843  838561493054533637    https://pbs.twimg.com/media/C6MrOsEXQAENOds.jpg   
230   670408998013820928    https://pbs.twimg.com/media/CU3FbQgVAAACdCQ.jpg   

      img_num                              p1   p1_conf  p1_dog  \
274         1                        Pekinese  0.609853    True   
521         1                     feather_boa  0.424106   False   
677         1                        keeshond  0.508951    True   
1777        1              Bedlington_terrier  0.392535    True   
1304        1                            chow  0.194773    True   
827         1                        sea_lion  0.519811   False   
945         1                         Samoyed  0.979206    True   
289         1             African_hunting_dog  0.733025   False   
304         1                Lakeland_terrier  0.428275    True   
1247        1        Chesapeake_Bay_retriever  0.804363    True   
938         2                    window_shade  0.909533   False   
1586        1  American_Staffordshire_terrier  0.489845    True   
990         1                           agama  0.172087   False   
1843        1                          kelpie  0.216562    True   
230         1                  ping-pong_ball  0.999945   False   

                           p2   p2_conf  p2_dog                         p3  \
274               Persian_cat  0.265442   False           Japanese_spaniel   
521         Yorkshire_terrier  0.073144    True          Shetland_sheepdog   
677                      chow  0.442016    True            German_shepherd   
1777       Labrador_retriever  0.089022    True                    clumber   
1304                  monitor  0.102305   False             Siberian_husky   
827               Siamese_cat  0.290971   False        black-footed_ferret   
945                Pomeranian  0.007185    True                 Arctic_fox   
289                      plow  0.119377   False         Scottish_deerhound   
304   wire-haired_fox_terrier  0.111472    True                 toy_poodle   
1247               Weimaraner  0.054431    True         Labrador_retriever   
938             window_screen  0.011427   False                      brass   
1586                Chihuahua  0.305760    True  Staffordshire_bullterrier   
990              Gila_monster  0.126978   False                 lumbermill   
1843                  doormat  0.139994   False                  dalmatian   
230               tennis_ball  0.000018   False                     racket   

       p3_conf  p3_dog  
274   0.027460    True  
521   0.057598    True  
677   0.013206    True  
1777  0.081800    True  
1304  0.086855    True  
827   0.039967   False  
945   0.006438   False  
289   0.026983    True  
304   0.105016    True  
1247  0.043268    True  
938   0.008882   False  
1586  0.072799    True  
990   0.050400   False  
1843  0.132820    True  
230   0.000015   False

In [190]:
tweet_json_df.head()

tweet_id  retweet_count  favorite_count
0  892420643555336193           6969           33693
1  892177421306343426           5272           29218
2  891815181378084864           3464           21974
3  891689557279858688           7191           36785
4  891327558926688256           7715           35178

In [191]:
tweet_json_df.tail()

tweet_id  retweet_count  favorite_count
2322  666049248165822465             36              88
2323  666044226329800704            115             246
2324  666033412701032449             36             100
2325  666029285002620928             39             112
2326  666020888022790149            419            2282

### Programmatic Assessment
**Quality Issues**

`twitter_enhanced_df`
1. 'None' values in `name`,`doggo`, `floofer`, `pupper` and `puppo` should be represents as `NaN`
2. 78 records are replies and are not of interest to us. We need only dog ratings
3. 181 records are retweets which are basically a repeat of an initial Twitter ID.
4. `timestamp` is an object.

`img_predictions_df`
1. 324 records in `img_predictions_df` are wrong. The actual pictures either show a different animal or the neural network couldn't detect the dog due to the dogs not properly represented in the photo.

**Tidiness Issues**
1. Join `tweet_json_df` to `twitter_enhanced_df` on `tweet_id`
2. Drop unneccesary (not useful in our viz) columns in `twitter_enhanced_df`: [1,2,6,7,8] 
3. Dog stages are separated into 4 columns (`doggo`, `floofer`, `pupper`, `puppo`) instead of just one [`twitter_enhanced_df`]

In [193]:
twitter_enhanced_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   tweet_id                    2356 non-null   int64  
 1   in_reply_to_status_id       78 non-null     float64
 2   in_reply_to_user_id         78 non-null     float64
 3   timestamp                   2356 non-null   object 
 4   source                      2356 non-null   object 
 5   text                        2356 non-null   object 
 6   retweeted_status_id         181 non-null    float64
 7   retweeted_status_user_id    181 non-null    float64
 8   retweeted_status_timestamp  181 non-null    object 
 9   expanded_urls               2297 non-null   object 
 10  rating_numerator            2356 non-null   int64  
 11  rating_denominator          2356 non-null   int64  
 12  name                        2356 non-null   object 
 13  doggo                       2356 

In [205]:
twitter_enhanced_df[twitter_enhanced_df.in_reply_to_status_id.notnull()].head(15)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
30   886267009285017600           8.862664e+17         2.281182e+09   
55   881633300179243008           8.816070e+17         4.738443e+07   
64   879674319642796034           8.795538e+17         3.105441e+09   
113  870726314365509632           8.707262e+17         1.648776e+07   
148  863427515083354112           8.634256e+17         7.759620e+07   
149  863079547188785154           6.671522e+17         4.196984e+09   
179  857214891891077121           8.571567e+17         1.806710e+08   
184  856526610513747968           8.558181e+17         4.196984e+09   
186  856288084350160898           8.562860e+17         2.792810e+08   
188  855862651834028034           8.558616e+17         1.943518e+08   
189  855860136149123072           8.558585e+17         1.361572e+07   
218  850333567704068097           8.503288e+17         2.195506e+07   
228  848213670039564288           8.482121e+17         4.196984e+09   
234  847617282490613760           8.476062e+17         4.196984e+09   
251  844979544864018432           7.590995e+17         4.196984e+09   

                     timestamp  \
30   2017-07-15 16:51:35 +0000   
55   2017-07-02 21:58:53 +0000   
64   2017-06-27 12:14:36 +0000   
113  2017-06-02 19:38:25 +0000   
148  2017-05-13 16:15:35 +0000   
149  2017-05-12 17:12:53 +0000   
179  2017-04-26 12:48:51 +0000   
184  2017-04-24 15:13:52 +0000   
186  2017-04-23 23:26:03 +0000   
188  2017-04-22 19:15:32 +0000   
189  2017-04-22 19:05:32 +0000   
218  2017-04-07 13:04:55 +0000   
228  2017-04-01 16:41:12 +0000   
234  2017-03-31 01:11:22 +0000   
251  2017-03-23 18:29:57 +0000   

                                                source  \
30   <a href="http://twitter.com/download/iphone" r...   
55   <a href="http://twitter.com/download/iphone" r...   
64   <a href="http://twitter.com/download/iphone" r...   
113  <a href="http://twitter.com/download/iphone" r...   
148  <a href="http://twitter.com/download/iphone" r...   
149  <a href="http://twitter.com/download/iphone" r...   
179  <a href="http://twitter.com/download/iphone" r...   
184  <a href="http://twitter.com/download/iphone" r...   
186  <a href="http://twitter.com/download/iphone" r...   
188  <a href="http://twitter.com/download/iphone" r...   
189  <a href="http://twitter.com/download/iphone" r...   
218  <a href="http://twitter.com/download/iphone" r...   
228  <a href="http://twitter.com/download/iphone" r...   
234  <a href="http://twitter.com/download/iphone" r...   
251  <a href="http://twitter.com/download/iphone" r...   

                                                  text  retweeted_status_id  \
30   @NonWhiteHat @MayhewMayhem omg hello tanner yo...                  NaN   
55   @roushfenway These are good dogs but 17/10 is ...                  NaN   
64                     @RealKentMurphy 14/10 confirmed                  NaN   
113  @ComplicitOwl @ShopWeRateDogs &gt;10/10 is res...                  NaN   
148  @Jack_Septic_Eye I'd need a few more pics to p...                  NaN   
149  Ladies and gentlemen... I found Pipsy. He may ...                  NaN   
179                       @Marc_IRL pixelated af 12/10                  NaN   
184  THIS IS CHARLIE, MARK. HE DID JUST WANT TO SAY...                  NaN   
186  @xianmcguire @Jenna_Marbles Kardashians wouldn...                  NaN   
188  @dhmontgomery We also gave snoop dogg a 420/10...                  NaN   
189  @s8n You tried very hard to portray this good ...                  NaN   
218  @markhoppus MARK THAT DOG HAS SEEN AND EXPERIE...                  NaN   
228  Jerry just apuppologized to me. He said there ...                  NaN   
234  .@breaannanicolee PUPDATE: Cannon has a heart ...                  NaN   
251  PUPDATE: I'm proud to announce that Toby is 23...                  NaN   

     retweeted_status_user_id retweeted_status_timestamp  \
30                        NaN                        NaN   
55                 

In [217]:
twitter_enhanced_df[twitter_enhanced_df.retweeted_status_id.notnull()].head()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
19  888202515573088257                    NaN                  NaN   
32  886054160059072513                    NaN                  NaN   
36  885311592912609280                    NaN                  NaN   
68  879130579576475649                    NaN                  NaN   
73  878404777348136964                    NaN                  NaN   

                    timestamp  \
19  2017-07-21 01:02:36 +0000   
32  2017-07-15 02:45:48 +0000   
36  2017-07-13 01:35:06 +0000   
68  2017-06-26 00:13:58 +0000   
73  2017-06-24 00:09:53 +0000   

                                               source  \
19  <a href="http://twitter.com/download/iphone" r...   
32  <a href="http://twitter.com/download/iphone" r...   
36  <a href="http://twitter.com/download/iphone" r...   
68  <a href="http://twitter.com/download/iphone" r...   
73  <a href="http://twitter.com/download/iphone" r...   

                                                 text  retweeted_status_id  \
19  RT @dog_rates: This is Canela. She attempted s...         8.874740e+17   
32  RT @Athletics: 12/10 #BATP https://t.co/WxwJmv...         8.860537e+17   
36  RT @dog_rates: This is Lilly. She just paralle...         8.305833e+17   
68  RT @dog_rates: This is Emmy. She was adopted t...         8.780576e+17   
73  RT @dog_rates: Meet Shadow. In an attempt to r...         8.782815e+17   

    retweeted_status_user_id retweeted_status_timestamp  \
19              4.196984e+09  2017-07-19 00:47:34 +0000   
32              1.960740e+07  2017-07-15 02:44:07 +0000   
36              4.196984e+09  2017-02-12 01:04:29 +0000   
68              4.196984e+09  2017-06-23 01:10:23 +0000   
73              4.196984e+09  2017-06-23 16:00:04 +0000   

                                        expanded_urls  rating_numerator  \
19  https://twitter.com/dog_rates/status/887473957...                13   
32  https://twitter.com/dog_rates/status/886053434...                12   
36  https://twitter.com/dog_rates/status/830583320...                13   
68  https://twitter.com/dog_rates/status/878057613...                14   
73  https://www.gofundme.com/3yd6y1c,https://twitt...                13   

    rating_denominator    name doggo floofer pupper puppo  
19                  10  Canela  None    None   None  None  
32                  10    None  None    None   None  None  
36                  10   Lilly  None    None   None  None  
68                  10    Emmy  None    None   None  None  
73                  10  Shadow  None    None   None  None

In [218]:
twitter_enhanced_df.describe()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
count  2.356000e+03           7.800000e+01         7.800000e+01   
mean   7.427716e+17           7.455079e+17         2.014171e+16   
std    6.856705e+16           7.582492e+16         1.252797e+17   
min    6.660209e+17           6.658147e+17         1.185634e+07   
25%    6.783989e+17           6.757419e+17         3.086374e+08   
50%    7.196279e+17           7.038708e+17         4.196984e+09   
75%    7.993373e+17           8.257804e+17         4.196984e+09   
max    8.924206e+17           8.862664e+17         8.405479e+17   

       retweeted_status_id  retweeted_status_user_id  rating_numerator  \
count         1.810000e+02              1.810000e+02       2356.000000   
mean          7.720400e+17              1.241698e+16         13.126486   
std           6.236928e+16              9.599254e+16         45.876648   
min           6.661041e+17              7.832140e+05          0.000000   
25%           7.186315e+17              4.196984e+09         10.000000   
50%           7.804657e+17              4.196984e+09         11.000000   
75%           8.203146e+17              4.196984e+09         12.000000   
max           8.874740e+17              7.874618e+17       1776.000000   

       rating_denominator  
count         2356.000000  
mean            10.455433  
std              6.745237  
min              0.000000  
25%             10.000000  
50%             10.000000  
75%             10.000000  
max            170.000000

In [220]:
twitter_enhanced_df[twitter_enhanced_df.rating_numerator > 100][['tweet_id','text','rating_numerator','rating_denominator','name']]

tweet_id                                               text  \
188   855862651834028034  @dhmontgomery We also gave snoop dogg a 420/10...   
189   855860136149123072  @s8n You tried very hard to portray this good ...   
290   838150277551247360                                 @markhoppus 182/10   
313   835246439529840640  @jonnysun @Lin_Manuel ok jomny I know you're e...   
902   758467244762497024  Why does this never happen at my front door......   
979   749981277374128128  This is Atticus. He's quite simply America af....   
1120  731156023742988288  Say hello to this unbelievably well behaved sq...   
1634  684225744407494656  Two sneaky puppers were not initially seen, mo...   
1635  684222868335505415  Someone help the girl is being mugged. Several...   
1779  677716515794329600  IT'S PUPPERGEDDON. Total of 144/120 ...I think...   
2074  670842764863651840  After so many requests... here you go.\n\nGood...   

      rating_numerator  rating_denominator     name  
188                420                  10     None  
189                666                  10     None  
290                182                  10     None  
313                960                   0     None  
902                165                 150     None  
979               1776                  10  Atticus  
1120               204                 170     this  
1634               143                 130     None  
1635               121                 110     None  
1779               144                 120     None  
2074               420                  10     None

Looks like the unusually high numerators and denominators are actually valid. Those are some good doggos 🤣

In [221]:
img_predictions_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   tweet_id  2075 non-null   int64  
 1   jpg_url   2075 non-null   object 
 2   img_num   2075 non-null   int64  
 3   p1        2075 non-null   object 
 4   p1_conf   2075 non-null   float64
 5   p1_dog    2075 non-null   bool   
 6   p2        2075 non-null   object 
 7   p2_conf   2075 non-null   float64
 8   p2_dog    2075 non-null   bool   
 9   p3        2075 non-null   object 
 10  p3_conf   2075 non-null   float64
 11  p3_dog    2075 non-null   bool   
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [222]:
img_predictions_df.describe()

tweet_id      img_num      p1_conf       p2_conf       p3_conf
count  2.075000e+03  2075.000000  2075.000000  2.075000e+03  2.075000e+03
mean   7.384514e+17     1.203855     0.594548  1.345886e-01  6.032417e-02
std    6.785203e+16     0.561875     0.271174  1.006657e-01  5.090593e-02
min    6.660209e+17     1.000000     0.044333  1.011300e-08  1.740170e-10
25%    6.764835e+17     1.000000     0.364412  5.388625e-02  1.622240e-02
50%    7.119988e+17     1.000000     0.588230  1.181810e-01  4.944380e-02
75%    7.932034e+17     1.000000     0.843855  1.955655e-01  9.180755e-02
max    8.924206e+17     4.000000     1.000000  4.880140e-01  2.734190e-01

In [249]:
img_predictions_df[(img_predictions_df.p1_dog + img_predictions_df.p2_dog + img_predictions_df.p3_dog) == 0]

tweet_id                                            jpg_url  \
6     666051853826850816    https://pbs.twimg.com/media/CT5KoJ1WoAAJash.jpg   
17    666104133288665088    https://pbs.twimg.com/media/CT56LSZWoAAlJj2.jpg   
18    666268910803644416    https://pbs.twimg.com/media/CT8QCd1WEAADXws.jpg   
21    666293911632134144    https://pbs.twimg.com/media/CT8mx7KW4AEQu8N.jpg   
25    666362758909284353    https://pbs.twimg.com/media/CT9lXGsUcAAyUFt.jpg   
...                  ...                                                ...   
2021  880935762899988482    https://pbs.twimg.com/media/DDm2Z5aXUAEDS2u.jpg   
2022  881268444196462592    https://pbs.twimg.com/media/DDrk-f9WAAI-WQv.jpg   
2046  886680336477933568    https://pbs.twimg.com/media/DE4fEDzWAAAyHMM.jpg   
2052  887517139158093824  https://pbs.twimg.com/ext_tw_video_thumb/88751...   
2074  892420643555336193    https://pbs.twimg.com/media/DGKD1-bXoAAIAUK.jpg   

      img_num                p1   p1_conf  p1_dog               p2   p2_conf  \
6           1        box_turtle  0.933012   False       mud_turtle  0.045885   
17          1               hen  0.965932   False             cock  0.033919   
18          1  desktop_computer  0.086502   False             desk  0.085547   
21          1  three-toed_sloth  0.914671   False            otter  0.015250   
25          1        guinea_pig  0.996496   False            skunk  0.002402   
...       ...               ...       ...     ...              ...       ...   
2021        1       street_sign  0.251801   False         umbrella  0.115123   
2022        1            tusker  0.473303   False  Indian_elephant  0.245646   
2046        1       convertible  0.738995   False       sports_car  0.139952   
2052        1         limousine  0.130432   False        tow_truck  0.029175   
2074        1            orange  0.097049   False            bagel  0.085851   

      p2_dog              p3   p3_conf  p3_dog  
6      False        terrapin  0.017885   False  
17     False       partridge  0.000052   False  
18     False        bookcase  0.079480   False  
21     False  great_grey_owl  0.013207   False  
25     False         hamster  0.000461   False  
...      ...             ...       ...     ...  
2021   False   traffic_light  0.069534   False  
2022   False            ibex  0.055661   False  
2046   False       car_wheel  0.044173   False  
2052   False   shopping_cart  0.026321   False  
2074   False          banana  0.076110   False  

[324 rows x 12 columns]

In [223]:
tweet_json_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2327 entries, 0 to 2326
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype
---  ------          --------------  -----
 0   tweet_id        2327 non-null   int64
 1   retweet_count   2327 non-null   int64
 2   favorite_count  2327 non-null   int64
dtypes: int64(3)
memory usage: 54.7 KB


In [224]:
tweet_json_df.describe()

tweet_id  retweet_count  favorite_count
count  2.327000e+03    2327.000000     2327.000000
mean   7.417930e+17    2458.667383     7026.269875
std    6.820795e+16    4163.681864    10919.212757
min    6.660209e+17       1.000000        0.000000
25%    6.781394e+17     492.500000     1220.500000
50%    7.178418e+17    1144.000000     3037.000000
75%    7.986547e+17    2844.500000     8565.000000
max    8.924206e+17   70330.000000   144246.000000

### Issues Found

#### Quality Issues

**`twitter_enhanced_df`**
1.  Some names in `name` column don't have the correct name
2.  Some rows in  have a 'None' value for `doggo`, `floofer`, `pupper` and `puppo` columns.
1. 'None' values in `name`,`doggo`, `floofer`, `pupper` and `puppo` should be represents as `NaN`
3. `in_reply_to_status_id`, `in_reply_to_user_id`, `retweeted_status_id`, `retweeted_status_user_id` and `retweeted_status_timestamp` have NaN values.
2. 78 records are replies and are not of interest to us. We need only dog ratings
3. 181 records are retweets which are basically a repeat of an initial Twitter ID.
4. `timestamp` is an object type.

**`img_predictions_df`**
1. For some Twitter IDs, `p1_dog`, `p2_dog` and `p3_dog` are all False indicating that there's no correct dog breed prediction for those IDs.
1. 324 records in `img_predictions_df` are wrong. The actual pictures either show a different animal or the neural network couldn't detect the dog due to the dogs not properly represented in the photo.
2. Names of dog breeds in `p1`, `p2` and `p3` are in lowercases.

**`tweet_json_df`**
1. 29 tweet IDs from `twitter_enhanced_df` don't have a record here.


#### Tidiness Issues
1. Join **`tweet_json_df`** to **`twitter_enhanced_df`** on `tweet_id`
1. **`img_predictions_df`** have 3 dog breeds. Only one with the %confidence is needed.
1. Join **`img_predictions_df`** to **`twitter_enhanced_df`** on `tweet_id`.
2. Drop unneccesary (not useful in our viz) columns in **`twitter_enhanced_df`**: [1,2,6,7,8] 
3. Dog stages are separated into 4 columns (`doggo`, `floofer`, `pupper`, `puppo`) instead of just one [**`twitter_enhanced_df`**]

## Cleaning Data